iree/compiler/Codegen/SPIRV/AdrenoConfig.cpp

```diff
@@ -0,0 +1,132 @@
+  // Y/X dimension, with each thread in a workgroup handling multiple vector
+  // elements. We start from the best (X, Y) and the tiling sizes for (M, N, K)
+  // and try different configurations by scaling them down until we find a
+  // configuration that can perfectly tile the input matmul.
+
+  const int64_t bestX = 32, bestY = 2;
+  const int64_t bestThreadM = 8, bestThreadN = 8, bestThreadK = 4;
+
+  int64_t residualThreads = bestX * bestY;
+  int64_t residualTilingFactor = (bestThreadM + bestThreadK) * bestThreadN;
+
+  SmallVector<int64_t, 3> workgroupSize(3, 1);        // (X, Y, Z)
+  SmallVector<int64_t, 4> workgroupTileSizes(3, 0);   // (M, N, K)
+  SmallVector<int64_t, 4> invocationTileSizes(3, 0);  // (M, N, K)
+
+  // Deduce the configuration for the N dimension. Start with the best workgroup
+  // X size, and reduce by a factor of two each time.
+  for (int64_t x = bestX; x >= 2; x >>= 1) {
+    int64_t chosenTileSize = 0;
+    // Also scale the thread tiling size by a factor of two each time. Don't go
+    // below four as we want vector load.
+    for (int64_t t = bestThreadN; t >= 4; t >>= 1) {
+      if (dimN % (x * t) == 0) {
+        chosenTileSize = t;
+        break;
+      }
+    }
+    if (chosenTileSize) {
+      workgroupSize[0] = x;
+      workgroupTileSizes[1] = x * chosenTileSize;
+      invocationTileSizes[1] = chosenTileSize;
+      residualThreads /= x;
+      assert(residualTilingFactor % chosenTileSize == 0);
+      residualTilingFactor /= chosenTileSize;
+      break;
+    }
+  }
+  if (workgroupTileSizes[1] == 0) return success();
+
+  // Deduce the configuration for the M dimension. Start with the best workgroup
+  // Y size, and reduce by a factor of two each time.
+  for (int64_t y = residualThreads; y >= 1; y >>= 1) {
+    int64_t chosenTileSize = 0;
+    // Reduce the thread tiling size by one each time. We read one row each
+    // time; so it's fine to not be some power of two here.
+    for (int64_t t = bestThreadM; t >= 1; --t) {
+      if (dimM % (y * t) == 0) {
+        chosenTileSize = t;
+        break;
+      }
+    }
+    if (chosenTileSize) {
+      workgroupSize[1] = y;
+      workgroupTileSizes[0] = y * chosenTileSize;
+      invocationTileSizes[0] = chosenTileSize;
+      assert(residualTilingFactor > chosenTileSize);
+      residualTilingFactor -= chosenTileSize;
+      break;
+    }
+  }
+  if (workgroupTileSizes[0] == 0) return success();
+
+  // Deduce the configuration for the K dimension. We need some power of two
+  // here so that we can do vector load.
+  for (int64_t t = llvm::PowerOf2Floor(residualTilingFactor); t >= 1; t >>= 1) {
+    if (dimK % t == 0) {
+      workgroupTileSizes[2] = invocationTileSizes[2] = t;
+      break;
+    }
+  }
+
+  auto pipeline = IREE::HAL::DispatchLoweringPassPipeline::SPIRVVectorize;
+  TileSizesListType tileSizes;
+  tileSizes.push_back(workgroupTileSizes);
+  tileSizes.emplace_back();
```

Comment:
We seem to be doing this everywhere. Not needed now, but either we dont need this level of tiling, or we need to separate these out.

Reply:
Noted. I'll just leave this as is for now to reduce churn. Will come to it after configurations for important architectures are fleshed out to a reasonable extent.

Reply:
Makes sense, this is definitely something we should be able to remove once with dynamic pipeline. It should be done separately though